packages/arb-ts/src/lib/message/L2ToL1Message.ts

```diff
@@ -0,0 +1,374 @@
+    // CHRIS: null check? this shouldn't be possible
+    // CHRIS: disable linting by just using the batchnumber here
+    batchNumber
+    return network.ethBridge!.outbox
+  }
+
+  public async getL2ToL1Messages<T extends SignerOrProvider>(
+    l1SignerOrProvider: T
+  ): Promise<L2ToL1MessageReaderOrWriter<T>[]>
+  public async getL2ToL1Messages<T extends SignerOrProvider>(
+    l1SignerOrProvider: T
+  ): Promise<L2ToL1MessageReader[] | L2ToL1MessageWriter[]> {
+    const provider = SignerOrProviderUtils.getProvider(l1SignerOrProvider)
+    if (!provider) throw new Error('Signer not connected to provider.')
+
+    const providerNetwork = await provider.getNetwork()
+    const arbNetwork = networks[providerNetwork.chainId]
+    return this.getL2ToL1Events().map(log => {
+      const outboxAddr = this.getOutboxAddr(arbNetwork, log.batchNumber)
+
+      return L2ToL1Message.fromBatchNumber(
+        l1SignerOrProvider,
+        outboxAddr,
+        log.batchNumber,
+        log.indexInBatch
+      )
+    })
+  }
+}
+
+/**
+ * Conditional type for Signer or Provider. If T is of type Provider
+ * then L2ToL1MessageReaderOrWriter<T> will be of type L2ToL1MessageReader.
+ * If T is of type Signer then L2ToL1MessageReaderOrWriter<T> will be of
+ * type L2ToL1MessageWriter.
+ */
+export type L2ToL1MessageReaderOrWriter<
+  T extends SignerOrProvider
+> = T extends Provider ? L2ToL1MessageReader : L2ToL1MessageWriter
+
+export class L2ToL1Message {
+  public static fromBatchNumber<T extends SignerOrProvider>(
+    l1SignerOrProvider: T,
+    outboxAddress: string,
+    batchNumber: BigNumber,
+    indexInBatch: BigNumber
+  ): L2ToL1MessageReaderOrWriter<T>
+  public static fromBatchNumber<T extends SignerOrProvider>(
+    l1SignerOrProvider: T,
+    outboxAddress: string,
+    batchNumber: BigNumber,
+    indexInBatch: BigNumber
+  ): L2ToL1MessageReader | L2ToL1MessageWriter {
+    return l1SignerOrProvider instanceof Provider
+      ? new L2ToL1MessageReader(
+          l1SignerOrProvider,
+          outboxAddress,
+          batchNumber,
+          indexInBatch
+        )
+      : new L2ToL1MessageWriter(
+          l1SignerOrProvider,
+          outboxAddress,
+          batchNumber,
+          indexInBatch
+        )
+  }
+
+  // CHRIS: consider what else to move onto the base
+  public static async getL2ToL1MessageLogs(
+    l2Provider: Provider,
+    filter: Filter
+  ) {
+    const iface = ArbSys__factory.createInterface()
+    const event = iface.getEvent('L2ToL1Transaction')
+    const eventTopic = iface.getEventTopic(event)
+
+    const topics = filter.topics ? filter.topics : []
+    // CHRIS: keep the warn?
+    // if (!filter.fromBlock && !filter.toBlock)
+    // console.warn('Attempting to query from 0 to block latest')
+    const logs = await l2Provider.getLogs({
+      address: ARB_SYS_ADDRESS,
+      topics: [eventTopic, ...topics],
+      fromBlock: filter.fromBlock || 0,
```

Comment:
I think these default to 0 / latest if undefined anyway, not sure why I did this

Reply:
Probably just having the param be a special thing that's just {fromBlock, toBlock, topics} makes more sense, since the address fields won't take effect anyway (I think @fredlacs has pointed this out)

Reply:
I've updated this in the asset-bridger branch